RFS/src/main/java/com/rfs/common/OpenSearchClient.java

```diff
@@ -0,0 +1,130 @@
+package com.rfs.common;
+
+import java.net.HttpURLConnection;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+import reactor.core.publisher.Mono;
+
+public class OpenSearchClient {
+    private static final Logger logger = LogManager.getLogger(OpenSearchClient.class);
+
+    public static class BulkResponse extends RestClient.Response {
+        public BulkResponse(int responseCode, String responseBody, String responseMessage) {
+            super(responseCode, responseBody, responseMessage);
+        }
+
+        public boolean hasBadStatusCode() {
+            return !(code == HttpURLConnection.HTTP_OK || code == HttpURLConnection.HTTP_CREATED);
+        }
+
+        public boolean hasFailedOperations() {
+            // The OpenSearch Bulk API response body is JSON and contains a top-level "errors" field that indicates
+            // whether any of the individual operations in the bulk request failed.  Rather than marshalling the entire
+            // response as JSON, just check for the string value.
+            return body.contains("\"errors\":true");
+        }
+
+        public String getFailureMessage() {
+            String failureMessage;
+            if (hasBadStatusCode()) {
+                failureMessage = "Bulk request failed.  Status code: " + code + ", Response body: " + body;
+            } else {
+                failureMessage = "Bulk request succeeded, but some operations failed.  Response body: " + body;
+            }
+
+            return failureMessage;
+        }
+    }
+
+    public final ConnectionDetails connectionDetails;
+    private final RestClient client;
+
+    public OpenSearchClient(ConnectionDetails connectionDetails) {
+        this.connectionDetails = connectionDetails;
+        this.client = new RestClient(connectionDetails);
+    }
+
+    /*
+     * Create a legacy template if it does not already exist; return true if created, false otherwise.
+     */
+    public boolean createLegacyTemplateIdempotent(String templateName, ObjectNode settings){
```

Comment:
Design Nit-pick: Separate use of idempotent and non-idempotent functionality with intermediate classes so you call looks like `client.idempotent().createLegacyTemplate(...)` vs `client.stateful().createSnapshot(...)`

Along these lines, expose the `client.idempotent()` interface at certain levels of the project to be really sure you aren't doing the wrong thing.

Reply:
Good suggestion; we'll see how the code develops.  This isn't the only thing in the class I'm not feeling confident in so I expect there to be continuing changes to it.